loki/transform/transform_array_indexing.py

```diff
@@ -496,3 +497,46 @@ def demote_variables(routine, variable_names, dimensions):
     routine.spec = Transformer(decl_map).visit(routine.spec)
 
     info(f'[Loki::Transform] Demoted variables in {routine.name}: {", ".join(variable_names)}')
+
+
+def flatten_arrays(routine, order='F', start_index=1):
+    """
+    Flatten arrays, converting multi-dimensional arrays to
+    one-dimensional arrays.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine in which the variables should be promoted.
+    order : str
+        Assume Fortran (F) vs. C memory/array order.
+    start_index : int
+        Assume array indexing starts with `start_index`.
+    """
+    def new_dims(dim, shape):
+        if len(dim) > 1:
+            assert not isinstance(shape[-1], sym.RangeIndex)
+            _dim = [sym.Sum((dim[-2], sym.Product((shape[-2], dim[-1] - start_index))))]
+            new_dim = dim[:-2]
+            new_dim.extend(_dim)
+            return new_dims(new_dim, shape[:-1])
+        return as_tuple(dim)
+
+    assert order in ['F', 'C']
+    if order == 'C':
+        array_map = {
+            var: var.clone(dimensions=new_dims(list(var.dimensions)[::-1], list(var.shape)[::-1]))
+            for var in FindVariables().visit(routine.body)
+                if isinstance(var, sym.Array) and var.shape and len(var.shape)
+        }
+    elif order == 'F':
+        array_map = {
+            var: var.clone(dimensions=new_dims(list(var.dimensions), list(var.shape)))
+            for var in FindVariables().visit(routine.body)
+                if isinstance(var, sym.Array) and var.shape and len(var.shape)
+        }
```

Comment:
To improve our error reporting, I'd suggest not to use `assert` here but instead throw an exception in an else branch:
```suggestion
    if order == 'C':
        array_map = {
            var: var.clone(dimensions=new_dims(list(var.dimensions)[::-1], list(var.shape)[::-1]))
            for var in FindVariables().visit(routine.body)
                if isinstance(var, sym.Array) and var.shape and len(var.shape)
        }
    elif order == 'F':
        array_map = {
            var: var.clone(dimensions=new_dims(list(var.dimensions), list(var.shape)))
            for var in FindVariables().visit(routine.body)
                if isinstance(var, sym.Array) and var.shape and len(var.shape)
        }
    else:
        raise ValueError(f'Unsupported array order "{order}")
```